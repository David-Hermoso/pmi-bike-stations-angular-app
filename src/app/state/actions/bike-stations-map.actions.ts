import {createAction, props} from '@ngrx/store';
import {BikeStation} from '../../models/bike-station';

export const bikeStationsMapActionsIds = {
  showBikeStationInfo: '[Bike stations map] show bike station info',
  closeBikeStationInfo: '[Bike stations info] close bike station info'
};

export const showBikeStationInfo = createAction(bikeStationsMapActionsIds.showBikeStationInfo,
  props<{ payload: BikeStation }>());

export const closeBikeStationInfo = createAction(bikeStationsMapActionsIds.closeBikeStationInfo);
