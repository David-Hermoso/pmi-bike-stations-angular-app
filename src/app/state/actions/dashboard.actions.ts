import {createAction, props} from '@ngrx/store';
import {BikeStation} from '../../dashboard';

export const dashboardActionsIds = {
  getBikeStationsStart: '[Dashboard] getBikeStations start',
  getBikeStationsSuccess: '[Dashboard] getBikeStations success',
  getBikeStationsError: '[Dashboard] getBikeStations error'
};

export const getBikeStationsStart = createAction(dashboardActionsIds.getBikeStationsStart);
export const getBikeStationsSuccess = createAction(dashboardActionsIds.getBikeStationsSuccess,
  props<{ bikeStations: BikeStation[] }>());
export const getBikeStationsError = createAction(dashboardActionsIds.getBikeStationsError);
