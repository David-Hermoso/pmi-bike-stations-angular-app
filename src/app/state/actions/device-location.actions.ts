import {createAction, props} from '@ngrx/store';

export const deviceLocationActionsIds = {
  request: '[Device location] request',
  success: '[Device location] success',
  error: '[Device location] error',
};

export const deviceLocationRequest = createAction(deviceLocationActionsIds.request);

export const deviceLocationSuccess = createAction(deviceLocationActionsIds.success,
  props<{ deviceLocation: any }>());
export const deviceLocationError = createAction(deviceLocationActionsIds.error);
