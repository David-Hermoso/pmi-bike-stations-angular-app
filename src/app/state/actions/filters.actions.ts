import {createAction, props} from '@ngrx/store';
import {Filters} from '../../models/filters';

export const filtersActionsIds = {
  filter: '[Filters] filter',
};

export const filter = createAction(filtersActionsIds.filter,
  props<{ payload: Filters }>());
