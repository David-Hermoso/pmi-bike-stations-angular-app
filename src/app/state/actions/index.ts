export * from './bike-stations-list.actions';
