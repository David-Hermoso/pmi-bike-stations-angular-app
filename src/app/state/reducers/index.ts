export * from './bike-stations-gallery.reducer';
