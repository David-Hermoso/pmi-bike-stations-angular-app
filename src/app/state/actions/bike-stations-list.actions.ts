import {createAction, props} from '@ngrx/store';
import {BikeStationsState} from '../reducers';

export const bikeStationsGalleryActionsIds = {
  getBikeStationsStart: '[Bike stations gallery] getBikeStations start',
  getBikeStationsSuccess: '[Bike stations gallery] getBikeStations success',
  getBikeStationsError: '[Bike stations gallery] getBikeStations error'
};

export const getBikeStationsStart = createAction(bikeStationsGalleryActionsIds.getBikeStationsStart);
export const getBikeStationsSuccess = createAction(bikeStationsGalleryActionsIds.getBikeStationsSuccess,
  props<BikeStationsState>());
export const getBikeStationsError = createAction(bikeStationsGalleryActionsIds.getBikeStationsError);
