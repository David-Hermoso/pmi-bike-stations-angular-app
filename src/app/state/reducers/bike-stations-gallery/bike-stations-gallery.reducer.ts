import {createReducer, createSelector, on} from '@ngrx/store';
import * as BikeStationsListActions from '../../actions/bike-stations-list.actions';
import * as FiltersActions from '../../actions/filters.actions';
import {AppState} from '../../../models/app.state';
import {BikeStation} from '../../../models/bike-station';
import {Filters} from '../../../models/filters';

export interface BikeStationsState {
  stations: BikeStation[];
}

export interface BikeStationsGalleryState {
  bikeStations: BikeStationsState;
  filters: Filters;
}

const initialFiltersState: Filters = {
  search: ''
};

export const initialBikeStationsState: BikeStationsGalleryState = {
  bikeStations: {
    stations: []
  },
  filters: initialFiltersState
};

export const selectBikeStationsGalleryState = (appState: AppState) => appState.bikeStationsGallery;
export const selectFilters = createSelector(
  selectBikeStationsGalleryState,
  ({filters}: BikeStationsGalleryState) => filters
);

export const selectBikeStations = createSelector(
  selectBikeStationsGalleryState,
  selectFilters,
  ({bikeStations, filters}: BikeStationsGalleryState) => {
    return filterBikeStations(bikeStations.stations, filters);
  }
);

export const bikeStationsGalleryReducer = createReducer(
  initialBikeStationsState,
  on(BikeStationsListActions.getBikeStationsSuccess,
    (currentState: BikeStationsGalleryState, {payload}) => ({
      ...currentState,
      bikeStations: {
        stations: payload.stations
      }
    })
  ),
  on(FiltersActions.filter, (currentState: BikeStationsGalleryState, {payload}) => ({
      ...currentState,
      filters: payload,
    })
  )
);

export const filterBikeStations = (bikeStations: BikeStation[], filters: Filters): BikeStation[] => {
  return bikeStations.filter(
    (bikeStation: BikeStation) => filterBikeStationByNameOrAddress(bikeStation.name, bikeStation.fullAddress, filters.search)
  );
};

export const filterBikeStationByNameOrAddress = (bikeStationName: string,
                                                 bikeStationAddress: string,
                                                 searchFilter: string) => {
  return (bikeStationName.toLowerCase().indexOf(searchFilter.toLowerCase()) > -1) ||
    bikeStationAddress.toLowerCase().indexOf(searchFilter.toLowerCase()) > -1;
};
