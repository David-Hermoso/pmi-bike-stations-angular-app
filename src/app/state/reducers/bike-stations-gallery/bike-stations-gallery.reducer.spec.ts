import * as fromBikeStationsGallery from './bike-stations-gallery.reducer';
import {
  BikeStationsGalleryState,
  filterBikeStationByNameOrAddress,
  filterBikeStations,
} from './bike-stations-gallery.reducer';
import {getBikeStationsSuccess} from '../../actions';
import {AppState} from '../../../models/app.state';
import {filter} from '../../actions/filters.actions';

describe('bikeStationsGalleryReducer', () => {
  describe('unknown action', () => {
    it('should return the default state', () => {
      const {initialBikeStationsState} = fromBikeStationsGallery;
      const action = {
        type: 'Unexistant action',
      };
      const state = fromBikeStationsGallery.bikeStationsGalleryReducer(initialBikeStationsState, action);

      expect(state).toBe(initialBikeStationsState);
    });
  });

  describe('getBikeStationsSuccess action', () => {
    it('should retrieve data of deployed apps and update the state', () => {
      const {initialBikeStationsState} = fromBikeStationsGallery;
      const newState: BikeStationsGalleryState = {
        bikeStations: {
          stations: []
        },
        filters: {
          search: ''
        }
      };

      const action = getBikeStationsSuccess({
        payload: newState.bikeStations
      });
      const state = fromBikeStationsGallery.bikeStationsGalleryReducer(initialBikeStationsState, action);

      expect(state).toEqual(newState);
    });
  });

  describe('filter action', () => {
    it('should return filtered bike stations list', () => {
      const {initialBikeStationsState} = fromBikeStationsGallery;
      const newState: BikeStationsGalleryState = {
        bikeStations: {
          stations: []
        },
        filters: {
          search: 'VALLDEMOSSA'
        }
      };

      const action = filter({
        payload: newState.filters
      });
      const state = fromBikeStationsGallery.bikeStationsGalleryReducer(initialBikeStationsState, action);

      expect(state).toEqual(newState);
    });
  });

  it('selectBikeStationsState: should return bike stations gallery state', () => {
    const appState: AppState = {
      bikeStationsGallery: {
        bikeStations: {
          stations: [
            {
              img: '',
              id: '16',
              name: 'PARC DE SA RIERA',
              lng: 2.644529343,
              lat: 39.5815637,
              fullAddress: 'Jesus, 21, 07010 Palma, Islas Baleares',
              street: 'Jesus',
              streetNumber: 21,
              cp: '07010',
              town: 'Palma',
              region: 'Illes Balears'
            },
            {
              img: null,
              id: '24',
              name: 'BLANQUERNA-SALLENT',
              lng: 2.651245594,
              lat: 39.57809894,
              fullAddress: 'Blanquerna, 2, 07003 Palma, Islas Baleares',
              street: 'Blanquerna',
              streetNumber: 2,
              cp: '07003',
              town: 'Palma',
              region: 'Illes Balears'
            }
          ]
        },
        filters: {
          search: ''
        }
      },
      selectedBikeStation: {
        selectedBikeStation: undefined
      },
    };
    expect(fromBikeStationsGallery.selectBikeStationsGalleryState(appState)).toStrictEqual(appState.bikeStationsGallery);
  });

  it('selectBikeStations: should return bike stations', () => {
    const appState: AppState = {
      bikeStationsGallery: {
        bikeStations: {
          stations: []
        },
        filters: {
          search: ''
        }
      },
      selectedBikeStation: {
        selectedBikeStation: undefined
      },
    };
    expect(fromBikeStationsGallery.selectBikeStations(appState)).toStrictEqual(appState.bikeStationsGallery.bikeStations.stations);
  });
});

it('filterBikeStationByName: should return truthy because bike station name matches filter name', () => {
  expect(
    filterBikeStationByNameOrAddress('Some bike station', 'Jesus, 21, 07010 Palma, Islas Baleares', 'jesus')
  ).toBeTruthy();
});

it('filterBikeStations: should return only one bike station', () => {
  expect(
    filterBikeStations([
      {
        img: null,
        id: '16',
        name: 'PARC DE SA RIERA',
        lng: 2.644529343,
        lat: 39.5815637,
        fullAddress: 'Jesus, 21, 07010 Palma, Islas Baleares',
        street: 'Jesus',
        streetNumber: 21,
        cp: '07010',
        town: 'Palma',
        region: 'Illes Balears'
      },
      {
        img: null,
        id: '24',
        name: 'BLANQUERNA-SALLENT',
        lng: 2.651245594,
        lat: 39.57809894,
        fullAddress: 'Blanquerna, 2, 07003 Palma, Islas Baleares',
        street: 'Blanquerna',
        streetNumber: 2,
        cp: '07003',
        town: 'Palma',
        region: 'Illes Balears'
      },
      {
        img: null,
        id: '77',
        name: 'SON COSTA - SON FORTEZA',
        lng: 2.6661,
        lat: 39.584114,
        fullAddress: 'Argelaga, 26, 07005 Palma, Illes Balears',
        street: 'Argelaga',
        streetNumber: 26,
        cp: '07005',
        town: 'Palma',
        region: 'Illes Balears'
      }
    ], {
      search: 'SON FORTEZA'
    })
  ).toMatchSnapshot();
});


