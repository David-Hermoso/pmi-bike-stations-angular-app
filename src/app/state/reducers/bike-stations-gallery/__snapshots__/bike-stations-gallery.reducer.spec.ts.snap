// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterBikeStations: should return only one bike station 1`] = `
Array [
  Object {
    "id": "77",
    "img": "001.jpeg",
    "kmToCityCenter": 4,
    "lat": 39.584114,
    "lng": 2.6661,
    "name": "SON COSTA - SON FORTEZA",
    "type": "anchor",
  },
]
`;
